src/components/molecules/TablesUserList/TablesUserList.tsx

```diff
@@ -44,169 +42,189 @@ const firestoreUpdate = (doc: string, update: any) => {
-  };
+  useEffect(() => {
+    userTable ? setSeatedAtTable(userTable) : setSeatedAtTable("");
+  }, [setSeatedAtTable, userTable]);
+
+  const isSeatedAtTable = seatedAtTable !== "";
 
   // @debt can we refactor this to make use of makeUpdateUserGridLocation ?
-  const takeSeat = (table: string) => {
-    if (!user) return;
-    const doc = `users/${user.uid}`;
-    const existingData = recentVenueUsers.find((u) => u.id === user.uid)?.data;
-
-    const update = {
-      data: {
-        ...existingData,
-        [venueName]: {
-          table,
-          videoRoom,
+  // @debt refactor this into api/* layer or similar?
+  const takeSeat = useCallback(
+    (table: string) => {
+      if (!user) return;
+
+      const doc = `users/${user.uid}`;
+      const existingData = recentVenueUsers.find((u) => u.id === user.uid)
+        ?.data;
+
+      const update = {
+        data: {
+          ...existingData,
+          [venueName]: {
+            table,
+            videoRoom,
+          },
         },
-      },
-    };
-    firestoreUpdate(doc, update);
-  };
+      };
+
+      firestoreUpdate(doc, update);
+    },
+    [recentVenueUsers, user, venueName, videoRoom]
+  );
+
+  const tableLocked = useCallback(
+    (table: string) => {
+      const areUsersAtTable = recentVenueUsers.some(
+        (user: User) => getUserExperience(venueName)(user)?.table === table
+      );
 
-  const usersAtOtherTables = [];
-  for (const table of tables) {
-    if (table.reference === seatedAtTable) {
-      continue;
-    }
-    usersAtOtherTables.push(...usersAtTables[table.reference]);
-  }
+      // Empty tables are never locked
+      if (!areUsersAtTable) return false;
+
+      // Locked state is in the experience record
+      return isTruthy(experience?.tables?.[table]?.locked);
+    },
+    [experience?.tables, recentVenueUsers, venueName]
+  );
+
+  const onAcceptJoinMessage = useCallback(
+    (table: string) => {
+      window.scrollTo(0, 0);
+      hideJoinMessage();
+      takeSeat(table);
+      setSeatedAtTable(table);
+    },
+    [hideJoinMessage, setSeatedAtTable, takeSeat]
+  );
+
+  const acceptJoiningTable = useCallback(
+    () => onAcceptJoinMessage(joiningTable),
+    [joiningTable, onAcceptJoinMessage]
+  );
+
+  const onJoinClicked = useCallback(
+    (table: string, locked: boolean, videoRoom: string) => {
+      if (locked) {
+        showLockedMessage();
+      } else {
+        setJoiningTable(table);
+        setVideoRoom(videoRoom);
+        joinMessage ? showJoinMessage() : onAcceptJoinMessage(table);
+      }
+    },
+    [joinMessage, onAcceptJoinMessage, showJoinMessage, showLockedMessage]
+  );
+
+  const renderedTables = useMemo(() => {
+    if (isSeatedAtTable) return;
+
+    return tables.map((table: Table, index: number) => (
+      <TableComponent
+        key={table.reference}
+        experienceName={venueName}
+        users={recentVenueUsers}
+        table={table}
+        tableLocked={tableLocked}
```

Comment:
[NB] This name is miss-leading, it indicates that this is a boolean but it's a function instead. This whole component and it's probably probably need to be refactored. Maybe add tech debt for it.

Reply:
Yup, this is part of the bundle of legacy tech debt that needs cleaning up. I think this is related to jazz bar from memory? That whole template/related is... a mess.. to say the least.

Reply:
> I think this is related to jazz bar from memory? That whole template/related is... a mess.. to say the least.

Yes.

Reply:
OOS